distributed/scheduler.py

```diff
@@ -722,104 +1239,148 @@ def update_graph(self, client=None, tasks=None, keys=None,
 
         if restrictions:
             # *restrictions* is a dict keying task ids to lists of
             # restriction specifications (either worker names or addresses)
-            worker_restrictions = defaultdict(set)
-            host_restrictions = defaultdict(set)
             for k, v in restrictions.items():
                 if v is None:
                     continue
+                ts = self.tasks.get(k)
+                if ts is None:
+                    continue
+                ts.host_restrictions = set()
+                ts.worker_restrictions = set()
                 for w in v:
                     try:
                         w = self.coerce_address(w)
                     except ValueError:
                         # Not a valid address, but perhaps it's a hostname
-                        host_restrictions[k].add(w)
+                        ts.host_restrictions.add(w)
                     else:
-                        worker_restrictions[k].add(w)
-
-            self.worker_restrictions.update(worker_restrictions)
-            self.host_restrictions.update(host_restrictions)
+                        ts.worker_restrictions.add(w)
 
             if loose_restrictions:
-                self.loose_restrictions |= set(loose_restrictions)
+                for k in loose_restrictions:
+                    ts = self.tasks[k]
+                    ts.loose_restrictions = True
 
         if resources:
-            self.resource_restrictions.update(resources)
+            for k, v in resources.items():
+                if v is None:
+                    continue
+                assert isinstance(v, dict)
+                ts = self.tasks.get(k)
+                if ts is None:
+                    continue
+                ts.resource_restrictions = v
 
         if retries:
-            self.retries.update(retries)
+            for k, v in retries.items():
+                assert isinstance(v, int)
+                ts = self.tasks.get(k)
+                if ts is None:
+                    continue
+                ts.retries = v
 
-        for key in sorted(touched | keys, key=self.priority.get):
-            if self.task_state[key] == 'released':
-                recommendations[key] = 'waiting'
+        # Compute recommendations
+        recommendations = OrderedDict()
 
-        for key in touched | keys:
-            for dep in self.dependencies[key]:
-                if dep in self.exceptions_blame:
-                    self.exceptions_blame[key] = self.exceptions_blame[dep]
+        for ts in sorted(runnables, key=operator.attrgetter('priority')):
+            if ts.state == 'released' and ts.run_spec:
+                recommendations[ts.key] = 'waiting'
+
+        for ts in touched_tasks:
+            for dts in ts.dependencies:
+                if dts.exception_blame:
+                    ts.exception_blame = dts.exception_blame
                     recommendations[key] = 'erred'
```

Comment:
Guessing `key` here should now be `ts.key`. Fixing in PR ( https://github.com/dask/distributed/pull/1900 ).